Gems/ROS2/Code/Source/ContactSensor/ROS2ContactSensorComponent.cpp

```diff
@@ -0,0 +1,203 @@
+#include "ROS2/Frame/ROS2FrameComponent.h"
+#include "ROS2/ROS2Bus.h"
+#include "ROS2/Sensor/ROS2SensorComponent.h"
+#include "ROS2/Utilities/ROS2Names.h"
+#include <cstddef>
+#include <geometry_msgs/msg/vector3.hpp>
+#include <geometry_msgs/msg/wrench.hpp>
+
+namespace ROS2
+{
+    ROS2ContactSensorComponent::ROS2ContactSensorComponent()
+    {
+        TopicConfiguration tc;
+        AZStd::string type = "gazebo_msgs::msg::ContactsState";
+        tc.m_type = type;
+        tc.m_topic = "contact_sensor";
+        m_sensorConfiguration.m_frequency = 15;
+        m_sensorConfiguration.m_publishersConfigurations.insert(AZStd::make_pair(AZStd::move(type), AZStd::move(tc)));
+    }
+
+    void ROS2ContactSensorComponent::Reflect(AZ::ReflectContext* context)
+    {
+        if (AZ::SerializeContext* serialize = azrtti_cast<AZ::SerializeContext*>(context))
+        {
+            serialize->Class<ROS2ContactSensorComponent, ROS2SensorComponent>()->Version(1);
+
+            if (AZ::EditContext* editContext = serialize->GetEditContext())
+            {
+                editContext->Class<ROS2ContactSensorComponent>("ROS2 Contact Sensor", "Contact detection controller")
+                    ->ClassElement(AZ::Edit::ClassElements::EditorData, "")
+                    ->Attribute(AZ::Edit::Attributes::Category, "ROS2")
+                    ->Attribute(AZ::Edit::Attributes::AppearsInAddComponentMenu, AZ_CRC_CE("Game"));
+            }
+        }
+    }
+
+    void ROS2ContactSensorComponent::GetRequiredServices(AZ::ComponentDescriptor::DependencyArrayType& required)
+    {
+        required.push_back(AZ_CRC_CE("PhysicsColliderService"));
+        required.push_back(AZ_CRC_CE("ROS2Frame"));
+    }
+
+    void ROS2ContactSensorComponent::Activate()
+    {
+        m_entityId = GetEntityId();
+        AZ::Entity* entity = nullptr;
+        AZ::ComponentApplicationBus::BroadcastResult(entity, &AZ::ComponentApplicationRequests::FindEntity, m_entityId);
+        m_entityName = entity->GetName();
+
+        auto ros2Node = ROS2Interface::Get()->GetNode();
+        AZ_Assert(m_sensorConfiguration.m_publishersConfigurations.size() == 1, "Invalid configuration of publishers for Contact sensor");
+        const auto publisherConfig = m_sensorConfiguration.m_publishersConfigurations["gazebo_msgs::msg::ContactsState"];
+        const auto fullTopic = ROS2Names::GetNamespacedName(GetNamespace(), publisherConfig.m_topic);
+        m_contactsPublisher = ros2Node->create_publisher<gazebo_msgs::msg::ContactsState>(fullTopic.data(), publisherConfig.GetQoS());
+
+        m_onCollisionBeginHandler = AzPhysics::SimulatedBodyEvents::OnCollisionBegin::Handler(
+            [this]([[maybe_unused]] AzPhysics::SimulatedBodyHandle bodyHandle, const AzPhysics::CollisionEvent& event)
+            {
+                AddNewContact(event);
+            });
+
+        m_onCollisionPersistHandler = AzPhysics::SimulatedBodyEvents::OnCollisionPersist::Handler(
```

Comment:
Why do you add a new contact if the collision persists?
Please check if `OnCollisionPersist` is needed in the first place.

Reply:
To ensure the most updated collision state (including contact points, forces, etc.) when publishing.

Reply:
Ok, right!